kvm: dump kvm_run area

masatake opened the pull request:
strae has `-e kvm=vcpu` already.
This pull request extends the option.
This pull request adds `-e kvm=vcpu+` option.
The new option is for dumping the kvm_run mmap'ed area.
```console
[yamato@dev64]~/var/strace/tests% ../src/strace -y -e ioctl -e kvm=vcpu+ ./ioctl_kvm_run_auxstr_vcpu
...
ioctl(5<anon_inode:kvm-vcpu:0>, KVM_RUN, 0) = 0 (KVM_EXIT_IO)
VCPU>{request_interrupt_window=0, immediate_exit=0, 
VCPU+ exit_reason=2 (KVM_EXIT_IO), ready_for_interrupt_injection=0, if_flag=0, flags=0, 
VCPU+ cr8=0000000000000000, apic_base=0000000000000000 => 0x000000fee00900, 
VCPU+ {io={direction=KVM_EXIT_IO_OUT, port=0x3f8, count=1, data_offset=0x00000000001000}}}
ioctl(5<anon_inode:kvm-vcpu:0>, KVM_RUN, 0) = 0 (KVM_EXIT_HLT)
VCPU>{request_interrupt_window=0, immediate_exit=0, 
VCPU+ exit_reason=5 (KVM_EXIT_HLT), ready_for_interrupt_injection=0, if_flag=0, flags=0, 
VCPU+ cr8=0000000000000000, apic_base=0x000000fee00900}
+++ exited with 0 +++
```

VCPU> and VCPU+ are the prefixes for the dump data.

ldv-alt commented:
> [yamato@dev64]~/var/strace/tests% ../src/strace -y -e ioctl -e kvm=vcpu+ ./ioctl_kvm_run_auxstr_vcpu
> ...
> ioctl(5<anon_inode:kvm-vcpu:0>, KVM_RUN, 0) = 0 (KVM_EXIT_IO)
> VCPU>{request_interrupt_window=0, immediate_exit=0, 
> VCPU+ exit_reason=2 (KVM_EXIT_IO), ready_for_interrupt_injection=0, if_flag=0, flags=0, 
> VCPU+ cr8=0000000000000000, apic_base=0000000000000000 => 0x000000fee00900, 
> VCPU+ {io={direction=KVM_EXIT_IO_OUT, port=0x3f8, count=1, data_offset=0x00000000001000}}}

Compared with other examples of strace's dump output (--read=set/--write=set), is it really a dump?
What I see here is a structure being printed in a very unusual way, with line prefixes and line breaks.
Could we come up with some nicer way of printing this information?

masatake commented on src/filter_qualify.c:
```diff
@@ -617,13 +617,22 @@ qualify_inject(const char *const str)
+		} else {
+		  wrong_kvm_qualifier:
+			error_msg("-e kvm=%s option needs"
+				  " Linux 4.16.0 or higher", str);
```
Updated ([243d022](https://github.com/strace/strace/pull/345/commits/243d0228506be468219f7a7222a9de438f402494)).

masatake commented:
I prepend a commit ([ad2a17f](https://github.com/strace/strace/pull/345/commits/ad2a17f264f7e38a00a8dddf09e39bf2aa007647)) to this pull request.

masatake commented:
I changed the prefix from "VCPU:%d> " to " VCPU:%d> ".

masatake commented:
Updated:

- renamed "dump_*" to "decode_*" in my changes.
- don't decode the kvm_run struct area when the last ioctl(KVM_RUN) fails.

@ldv-alt I would like to get your review again.

masatake commented:
It seems that net-yy-unix.test and linkat--secontext_mismatch.gen failed.

ldv-alt reviewed (commented):
Please correct the following part of the commit message:
```
(kvm_run_structure_decoder_init): Store the passed
the enum decode_kvm_run_structure_modes argument to
decode_kvm_run_structure.
```

For example,
```
(kvm_run_structure_decoder_init): Save
enum decode_kvm_run_structure_modes argument into
decode_kvm_run_structure.
```

ldv-alt commented on src/kvm.c:
```diff
@@ -351,14 +356,28 @@ kvm_ioctl_run_attach_auxstr(struct tcb *const tcp,
+					tcp->vcpu_entering
+						= xzalloc(sizeof(*tcp->vcpu_entering));
+				kvm_ioctl_run_umove(tcp, info,
+						    tcp->vcpu_entering);
```
What if kvm_ioctl_run_umove failed?  Wouldn't vcpu_entering contain garbage in that case?

ldv-alt commented on tests/ioctl_kvm_run_auxstr_vcpu.c:
```diff
@@ -12,20 +12,35 @@
 
 #if need_print_KVM_RUN
 
+#define need_print_KVM_RUN_MORE 1
```
I don't see why `need_print_KVM_RUN_MORE` is being defined in this file.

masatake replied:
need_print_KVM_RUN_MORE is redundant.I will remove this.

masatake commented:
@ldv-alt Thank you for reviewing. I've incorporated some of your comments into the commits.
I will continue to work on the remaining items.

masatake commented:
> Could you provide a rationale for this asm change, please?

I updated the commit log.

masatake commented:
@ldv-alt I've reflected all of your points. Could you review again?

ldv-alt reviewed (commented):
Please add a NEWS entry.

masatake commented:
```
* Added -e kvm=vcpu+ option for decoding kvm_run struct.
```

Updated.

masatake commented:
Thank you for reviewing and merging.
